sycl/doc/extensions/DeviceGlobal/SPV_INTEL_global_variable_decorations.asciidoc

```diff
@@ -0,0 +1,224 @@
+
+This extension requires SPIR-V 1.0.
+
+== Overview
+
+This extension adds decorations that can be applied to global (module scope)
+variables.  These decorations are mostly intended to help code generation for
+FPGA devices, but they must be recognized by any consumer that supports this
+extension.
+
+== Extension Name
+To use this extension within a SPIR-V module, the following *OpExtension* must
+be present in the module:
+
+----
+OpExtension "SPV_INTEL_global_variable_decorations"
+----
+
+== New Capabilities
+This extension introduces a new capability:
+
+----
+GlobalVariableDecorationsINTEL
+----
+
+== Token Number Assignments
+
+--
+[width="40%"]
+[cols="70%,30%"]
+[grid="rows"]
+|====
+|GlobalVariableDecorationsINTEL | 6146
+|HostAccessINTEL | 6147
+|InitModeINTEL | 6148
+|ImplementInCSRINTEL | 6149
+|====
+--
+
+== Modifications to the SPIR-V Specification, Version 1.5
+
+=== Decoration
+
+Modify Section 3.20, Decoration, adding these rows to the Decoration table:
+
+--
+[options="header"]
+|====
+2+^| Decoration 2+^| Extra Operands | Enabling Capabilities
+
+// --- ROW BREAK ---
+| 6147
+a|
+*HostAccessINTEL* +
+Only valid on global (module scope) *OpVariable*.
+
+The client API's execution environment may provide a way to access a global
+variable's value from the host system.  If it does, this decoration provides
+two pieces of information.  _Access_ is an assertion by the producer about the
+types of these accesses, which may allow the consumer to perform certain
+optimizations, especially for FPGA devices.  _Name_ is a name which the client
+API's execution environment may use to identify this variable.
+
+Legal values of _Access_:
+
+* 0 [Read] - The execution environment may read the variable from the host but
+  will never write it.  On an FPGA device, only a read memory port is exposed
+  by the generated IP core.
+* 1 [Write] - The execution environment may write the variable from the host
+  but will never read it.  On an FPGA device, only a write memory port is
+  exposed by the generated IP core.
+* 2 [Read/Write] - The execution environment may read or write the variable
+  from the host.  On an FPGA device, a read/write memory port is exposed by the
+  generated IP core.
+* 3 [None] - The execution environment may neither read nor write the variable
+  from the host.  On an FPGA device, no memory port is exposed by the generated
+  IP core.
+
+If a global *OpVariable* is not decorated with *HostAccessINTEL*, the default
+behavior is [Read/Write].
+| Literal Number +
+_Access_
+| Literal String +
+_Name_
+| *GlobalVariableDecorationsINTEL*
+
+// --- ROW BREAK ---
+| 6148
+a|
+*InitModeINTEL* +
+Only valid on global (module scope) *OpVariable*.
+
+This decoration only has an effect when the consumer is an FPGA or similar
+device.  The _Trigger_ value tells how the global variable should be
+initialized.
+
+Legal values of _Trigger_:
+
+* 0 [init on device reprogram] - Initialization is performed when the device is
```

Comment:
Please don't use the word "when" here.  This control does not change when initialization happens only how it is performed.

Reply:
I suggest using the same wording from the device global API specification:

> Initialization is performed by reprogramming the device. This may require more frequent reprogramming but may reduce area.
>
>  Initialization is performed by sending a reset signal to the device. This may increase area but may reduce reprogramming frequency.

Seem good?

Reply:
Yes, to your first change.  I'm a little wary of the second one because it seems overly prescriptive.  What if an implementation wants to implement initialization by launching dedicated initialization kernels for example?  I guess that would have the same performance/area tradeoff as "init on device reset" but it technically wouldn't be accomplished by sending a reset signal to the device, it would be accomplished by enqueuing a kernel.  This makes me think that the wording in the SYCL spec is also overly prescriptive.  @artemrad would do you think about dropping this sentence from the SYCL spec?

Reply:
I think we should keep the sentence. This property doesn't have an "off" value or at least behavior that is intuitive when this property is missing. The sentence make it explicit that something is going to happen related to this property whether you include it or not.

Reply:
What if weaken it then to something non-normative such as: "Note that the presence or absence of this decoration does not affect when initialization happens."?

Reply:
@GarveyJoe:

I interpreted @artemrad's comment as saying he does not want to remove this sentence from the [API spec](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc):

> If the init_mode property is not specified, the default behavior is equivalent to one of the values listed above, but the choice is implementation defined.

Replacing that sentence with the one you propose does not make sense to me.  I assume you must be interpreting @artemrad's comment differently?

Reply:
We discussed this offline and decided that the wording about the default behavior should just be less proscriptive.  I did this in f0d951288eaa4f2d2a0244b3907231fe77cc6540.

Note that I also updated the wording to say that **InitModeINTEL** is only allowed for an **OpVariable** that has an _Initializer_ operand.  I presume that this decoration does not make sense for variables with no initial value.